gulp/util/content.js

```diff
@@ -0,0 +1,204 @@
+/*
+ * 2. to convert markdown to html
+ * 3. handlebars is used to convert `{{ }}` placeholders
+ *    in markdown, html, to output
+ * 4. brings in our own shared `utils`
+ */
+var fs = require('fs');
+var path = require('path');
+var md = require('markdown-it')({ html: true }); /* 2 */
+var Handlebars = require('handlebars'); /* 3 */
+var utils = require('../util/template-utils.js'); /* 4 */
+
+require('events').EventEmitter.prototype._maxListeners = 100;
+
+function traverse(obj, str) {
+  return str.split(".").reduce(function(o, x) { return o[x] }, obj);
+}
+
+/*
+  generateContentAndTemplates()
+  =============
+  This function wraps some lookups and caching around otherwise repeated actions
+  within the run of the task returned.
+
+  In general, the purpose is to:
+  - take incoming Markdown files** and inject in dictionary variables
+  - render the post-processed Markdown in to HTML
+  - fetch the appropriate template (HTML)
+  - Inject in dictionary variables and feed the HTML **content** (from markdown)
+    in to the template.
+  - Return the final combined HTML through to the vinyl stream.
+
+  ** later, we want to accept incoming HTML partials as well
+     (not all pages will be Markdown based)
+
+  Returns: a gulp-friendly pipe task (function)
+*/
+function generateContentAndTemplates() {
+  var base, projectJSON, i18nJSON, hbsTemplates, LocalHandlebars;
+
+  /*
+   * cache variables and lookups used on subsequent runs of the pipe task:
+   *
+   * 1. `base` directory of project
+   * 2. `contentBase` is the root directory where the task is getting its content,
+   *     this is helpful later for processing out which i18n we're looking at
+   * 3. `projectJSON` is global, re-used across all languages
+   * 4. `i18nJSON` caches the template JSON for each language (avoids duplicated work)
+   * 5. `hbsTemplates` caches the handlebars FUNCTION for each template to save overhead
+   * 5. `LocalHandlebars` is a sandboxed version of Handlebars, avoids injecting
+         helpers and partials at a global scale.
+   */
+  base = path.resolve(__dirname, '..', '..'); /* 1 */
+  contentBase = path.resolve(base, 'content'); /* 2 */
+  projectJSON = require('../../source/project.js'); /* 3 */
+  i18nJSON = {}; /* 4 */
+  hbsTemplates = {}; /* 5 */
+  LocalHandlebars = Handlebars.create() /* 6 */
+
+  LocalHandlebars.registerPartial('current_download_links', `{{#project.current_version_downloads}}<a href="{{url}}">{{i18n "downloads" key}}</a>{{/project.current_version_downloads}}`)
+
+  LocalHandlebars.registerHelper('i18n', function() {
+    var scope, i18n_key, env, key, data, lang, result;
+
+    var args = Array.prototype.slice.call(arguments);
```

Comment:
to be fair, there isn't too much of a compelling case to make a copy here, it looks like we should just be able to access `arguments` since we are only using `.length` and `[knownIndex]`